### PR TITLE
Enhancement proposal for OpenShift IPI on Nutanix AOS with AHV

### PR DESCRIPTION
**Pull request by vnephologist**
This is initial enhancement request to support IPI installer for Nutanix AOS with AHV hypervisor.

### CODE REVIEW

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
+---
+title: nutanix-ahv-ipi
+authors:
+  - "@vnephologist"
+reviewers:
+  - "@makentenza"
+  - "@fabianofranz"
```
Comment (dhellmann) [inline enhancements/installer/nutanix-ahv-ipi.md:7]: @staebler Who from the installer team should we list as a reviewer?
@JoelSpeed Who from the cluster infrastructure team should we list as a reviewer?

Comment (elmiko) [reply enhancements/installer/nutanix-ahv-ipi.md:7]: you can list me as a reviewer, i'll help out

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+Red Hat and Nutanix have recently [announced](https://www.redhat.com/en/about/press-releases/red-hat-and-nutanix-announce-strategic-partnership-deliver-open-hybrid-multicloud-solutions) a strategic partnership to deliver open hybrid multicloud solutions. For OpenShift this means that it ​​becomes the preferred enterprise full-stack Kubernetes solution on Nutanix Cloud Platform with AHV.
+
+Today, OpenShift can be installed on the Nutanix AOS platform using the OpenShift platform [agnostic installer](https://github.com/nutanix/openshift/tree/main/docs/install/manual), but mutual customers have also requested a fully automated, integrated installer built using OpenShift 4's infrastructure level automation capabilities. This is particularly appealing for customers deploying in "hybrid cloud" scenarios where OpenShift should be deployed using common mechanisms across different clouds.
+
+### Goals
+
+The primary goal of Nutanix AHV IPI is to provide users with an easier path to running OpenShift 4 on Nutanix AOS with AHV.
+
+To achieve that goal, we will:
```
Comment (elmiko) [inline enhancements/installer/nutanix-ahv-ipi.md:41]: is there also a Nutanix [Cloud Controller Manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) to mention here?

**Review by elmiko — COMMENTED** [review]
just starting to review this, mostly makes sense but i have a feeling there will be more details coming?

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  ? SSH Public Key /Users/someone/.ssh/id_rsa.pub
+  ? Platform nutanix
+  ? Prism Central pc01.ocp.nutanix.com
+  ? Prism Central Username admin@ocp.nutanix.com
+  ? Prism Central Password [? for help] **********
+  ? Cluster (Prism Element) nx01.ocp.nutanix.com
+  ? Container default-container-70593
+  ? Subnet vm-network
+  ? Virtual IP Address for API 192.168.1.5
+  ? Virtual IP Address for Ingress 192.168.1.6
```
Comment (staebler) [inline enhancements/installer/nutanix-ahv-ipi.md:65]: This is a lot of questions to include in the survey. Let's consider whether there are reasonable defaults for any of these? For example, are there defaults for the VIP addresses?

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+- Implement Nutanix specific platform installer customizations:
+  
+  **Platform**
+  
+  - prismCentral (required string): Name or IP address of the Nutanix Prism Central instance to target.
+  - prismCentralUser (required string): Username for Prism Central access (Cluster Admin).
+  - prismCentralPass (required string): Password for Prism central access.
+  - cluster (required string): Nutanix cluster (Prism Element) where the OpenShift cluster will be created.
+  - container (required string): Nutanix container name for OS disk install.
+  - Subnet (required string): Nutanix subnet name (IPAM or DHCP Enabled).
```
Comment (staebler) [inline enhancements/installer/nutanix-ahv-ipi.md:80]:
```suggestion
  - subnet (required string): Nutanix subnet name (IPAM or DHCP Enabled).
```

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+Story 3
+As an OpenShift platform, I want OpenShift on Nutanix to be maintained and released like other supported OpenShift platforms and covered by CI tests.
+
+### Implementation Details/Notes/Constraints
+
+- Nutanix AHV supports VM guest customization via [cloud-init Config Drive v2](https://cloudinit.readthedocs.io/en/latest/topics/datasources/configdrive.html).
+
+- It will be assumed the customer utilizes a subnet configured with Nutanix IPAM or will provide a properly configured DHCP server that is available on the configured L2 network.
+
+- Nutanix does not provide networking services like DNS or load balancing. With this in mind, we require the ability to automate DNS and load balancing services internal to the cluster. The a solution similar to the baremetal-networking enhancement can be utilized.
```
Comment (staebler) [inline enhancements/installer/nutanix-ahv-ipi.md:104]:
```suggestion
- Nutanix does not provide networking services like DNS or load balancing. With this in mind, we require the ability to automate DNS and load balancing services internal to the cluster. A solution similar to the baremetal-networking enhancement can be utilized.
```

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+Today, OpenShift can be installed on the Nutanix AOS platform using the OpenShift platform [agnostic installer](https://github.com/nutanix/openshift/tree/main/docs/install/manual), but mutual customers have also requested a fully automated, integrated installer built using OpenShift 4's infrastructure level automation capabilities. This is particularly appealing for customers deploying in "hybrid cloud" scenarios where OpenShift should be deployed using common mechanisms across different clouds.
+
+### Goals
+
+The primary goal of Nutanix AHV IPI is to provide users with an easier path to running OpenShift 4 on Nutanix AOS with AHV.
+
+To achieve that goal, we will:
+- Enhance the installer, adding Nutanix AOS options.
+- Prepare Nutanix infrastructure using the [Nutanix Terraform (TF) provider](https://github.com/nutanix/terraform-provider-nutanix) for provisioning.
+- Integrate with the Nutanix cluster API provider [to-do] and enhance requisite operators to enable cluster functionality.
```
Comment (staebler) [inline enhancements/installer/nutanix-ahv-ipi.md:44]: When doing this integration, please consider how certificate trust bundles will be provided to the cluster API provider and the in-cluster operators when the Nutanix endpoints are secured with internal CAs.

File: enhancements/installer/nutanix-ahv-ipi.md (hunk 2 above)
Comment (vnephologist) [reply enhancements/installer/nutanix-ahv-ipi.md:41]: We discussed offline and will initially handle node lifecycle via MAPI and implement CCM with broader scope at a later date.

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+Story 3
+As an OpenShift platform, I want OpenShift on Nutanix to be maintained and released like other supported OpenShift platforms and covered by CI tests.
+
+### Implementation Details/Notes/Constraints
+
+- Nutanix AHV supports VM guest customization via [cloud-init Config Drive v2](https://cloudinit.readthedocs.io/en/latest/topics/datasources/configdrive.html).
+
+- It will be assumed the customer utilizes a subnet configured with Nutanix IPAM or will provide a properly configured DHCP server that is available on the configured L2 network.
+
+- Nutanix does not provide networking services like DNS or load balancing. With this in mind, we require the ability to automate DNS and load balancing services internal to the cluster. A solution similar to the baremetal-networking enhancement can be utilized.
```
Comment (rvanderp3) [inline enhancements/installer/nutanix-ahv-ipi.md:112]: From discussions, you might update this bullet point to reflect that Nutanix will be leveraging [on-prem infra](https://github.com/openshift/machine-config-operator/tree/master/templates/common/on-prem)

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+### Risks and Mitigations
+
+- The Nutanix Cluster API provider is a work-in-progress.  It is currently being tracked internally at Nutanix and expected to be open sourced.
+
+## Design Details
+
+- Cloud Credential Operator (CCO) will operate in manual mode. Possible migration to passthrough mode in future enhancement.
```
Comment (rvanderp3) [inline enhancements/installer/nutanix-ahv-ipi.md:120]: You might add 'Possible migration to passthrough mode in future enhancement' to the out-of-scope heading

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+- Cloud Controller Manager (CCM) is not currently implemented. Node lifecycle handled via MAPI until standard AOS/AHV CCM is developed.
+
+- CSI Operator already publised [here](https://catalog.redhat.com/software/containers/nutanix/nutanix-csi-operator/60dfad1364002f2e99660a86). Installer to deploy CSI and provision PV for use by Internal Registry.
+
+- Internal Registry to utilize Nutanix Volumes (block) storage by default. Post-install configuration on Nutanix Files or Objects to be documented.
+
+### Open Questions
+
+1. How should VM specifications be determined or specified?
```
Comment (rvanderp3) [inline enhancements/installer/nutanix-ahv-ipi.md:130]: See [Minimum resource requirements for cluster installation
](
https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-minimum-resource-requirements_installing-bare-metal)

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+1. Are there any minimum host requirements to enforce?
+
+1. Should the default install attempt to locate control plane VMs on different physical nodes?
+
+1. Are unit-level tests recommended (required?). If so, is there preferred tooling / test structure? This is in addition to the required CI E2E testing.
```
Comment (rvanderp3) [inline enhancements/installer/nutanix-ahv-ipi.md:136]: Unit tests are often written against the [testing package](https://pkg.go.dev/testing).  Unit tests are required where possible.

File: enhancements/installer/nutanix-ahv-ipi.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+- Implement Nutanix specific platform installer customizations:
+  
+  **Platform**
+  
+  - prismCentral (required string): Name or IP address of the Nutanix Prism Central instance to target.
+  - prismCentralUser (required string): Username for Prism Central access (Cluster Admin).
+  - prismCentralPass (required string): Password for Prism central access.
+  - cluster (required string): Nutanix cluster (Prism Element) where the OpenShift cluster will be created.
+  - container (required string): Nutanix container name for OS disk install.
+  - subnet (required string): Nutanix subnet name (IPAM or DHCP Enabled).
```
Comment (rvanderp3) [inline enhancements/installer/nutanix-ahv-ipi.md:88]: Would it be possible to add a sample 'platform' section of an install-config.yaml?

File: enhancements/installer/nutanix-ahv-ipi.md (hunk 8 above)
Comment (vnephologist) [reply enhancements/installer/nutanix-ahv-ipi.md:120]: It doesn't seem like it is desirable to even move to passthrough, so will remove entirely.

File: enhancements/installer/nutanix-ahv-ipi.md (hunk 11 above)
Comment (vnephologist) [reply enhancements/installer/nutanix-ahv-ipi.md:88]: Done

**Pull request closed without merging**